public/pages/workflow_detail/workspace_component/utils.ts

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Connection, ReactFlowInstance } from 'reactflow';
+import { IComponentInput } from '../../../../common';
+
+/**
+ * Collection of utility functions for the workspace component
+ */
+
+// Uses DOM elements to calculate where the handle should be placed
+// vertically on the ReactFlow component. offsetTop is the offset relative to the
+// parent element, and clientHeight is the element height including padding.
+// We can combine them to get the exact amount, in pixels.
+export function calculateHandlePosition(ref: any): number {
+  if (ref.current && ref.current.offsetTop && ref.current.clientHeight) {
+    return ref.current.offsetTop + ref.current.clientHeight / 2;
+  } else {
+    return 0;
+  }
+}
+
+// Validates that connections can only be made when the source and target classes align, and
+// that multiple connections to the same target handle are not allowed unless the input configuration
+// for that particular component allows for it.
+export function isValidConnection(
```

Comment:
quick question if we do here the validation if the source class aligns with target class, what is left for the onConnect method where we do validation as well?

Reply:
`onConnect` will handle any other updates we need to do after a connection is made, such as updating properties of the edge itself, or update properties of the node being connected potentially. `isValidConnection` is just for gatekeeping that a connection is allowed to be made or not.